File: renaissance-core/src/main/scala/org/renaissance/renaissance-benchmark.scala
package org.renaissance

import java.util.regex.Pattern
import scala.util.Failure
import scala.util.Success
import scala.util.Try

trait RenaissanceBenchmark {
  final def name: String = {
    val nameOfClass = this.getClass.getSimpleName
    val camelCaseName = if (nameOfClass.last == '$') nameOfClass.init else nameOfClass
    val pattern = Pattern.compile("([A-Za-z])([A-Z])")
    pattern.matcher(camelCaseName).replaceAll("$1-$2").toLowerCase
  }

  final def mainGroup: String = {
    val fullName = getClass.getName
    val simpleName = getClass.getSimpleName
    val packageName = fullName.substring(0, fullName.indexOf(simpleName)).init
    val groupName = packageName.substring(packageName.lastIndexOf('.') + 1)
    groupName
  }

  def defaultRepetitions = 20

  def description: String

  def initialRelease: Option[String] = None

  def finalRelease: Option[String] = None

  protected def setUpBeforeAll(c: Config) = {}

  protected def tearDownAfterAll(c: Config) = {}

  protected def beforeIteration(c: Config) = {}

  protected def afterIteration(c: Config) = {}

  final def runBenchmark(config: Config): Try[Unit] = {
    try {
      setUpBeforeAll(config)
      val factory = Policy.factories.getOrElse(config.policy, sys.error("Unknown policy."))
      val policy = factory.apply(this, config)
      for (plugin <- config.plugins) plugin.onStart(policy)
      try policy.execute()
      finally for (plugin <- config.plugins) plugin.onTermination(policy)
    } catch {
      case t: Throwable =>
        return Failure(t)
    } finally {
      tearDownAfterAll(config)
    }
  }

  protected def runIteration(config: Config): Unit

  def runIterationWithBeforeAndAfter(policy: Policy, config: Config): Long = {
    beforeIteration(config)

    for (plugin <- config.plugins) plugin.beforeIteration(policy)

    val start = System.nanoTime

    runIteration(config)

    val end = System.nanoTime
    val duration = end - start

    for (plugin <- config.plugins) plugin.afterIteration(policy, duration)

    afterIteration(config)

    duration
  }
}

case class Config(
  val benchmarkList: Seq[String] = Seq(),
  val repetitions: Int = -1,
  val plugins: Seq[Plugin] = Seq(),
  val policy: String = "fixed",
  val readme: Boolean = false
) {

  def withBenchmarkSpecification(v: String): Config =
    this.copy(benchmarkList = v.split(",").toSeq)

  def withPlugins(v: String): Config = {
    val pluginNames = v.split(",")
    val plugins = for (name <- pluginNames) yield {
      Class.forName(name).newInstance().asInstanceOf[Plugin]
    }
    this.copy(plugins = plugins)
  }
}

/** Base class for plugins that gather other metrics.
 *
 *  Subclasses must have a zero arguments constructor.
 */
trait Plugin {

  /** Called once after the plugin is created.
   */
  def onCreation(): Unit

  /** Called before a benchmark starts executing.
   */
  def onStart(policy: Policy): Unit = {}

  def beforeIteration(policy: Policy): Unit

  def afterIteration(policy: Policy, duration: Long): Unit

  /** Called after the benchmark is done executing.
   */
  def onTermination(policy: Policy): Unit = {}

  /** Called once after all benchmarks are done executing.
   */
  def onExit(): Unit = {}
}

/** The policy that executes the benchmark
 */
sealed trait Policy {
  def description: String

  /** The name of the benchmark that is currently executing.
   */
  def currentBenchmarkName: String = currentBenchmark.name

  /** The benchmark that is executing.
   */
  private[renaissance] def currentBenchmark: RenaissanceBenchmark

  /** Runs the benchmark multiple times.
   */
  private[renaissance] def execute(): Try[Unit]
}

object Policy {

  val factories = Map[String, (RenaissanceBenchmark, Config) => Policy](
    "fixed" -> { (benchmark, config) =>
      new FixedIterationsPolicy(benchmark, config)
    }
  )

  def descriptions: Map[String, String] = factories.map {
    case (name, factory) =>
      val policy = factory(new Dummy, new Config)
      (name, policy.description)
  }
}

/** Represents a run in which a fixed number of iterations are sequentially executed.
 */
final class FixedIterationsPolicy(
  private[renaissance] val currentBenchmark: RenaissanceBenchmark,
  private val config: Config
) extends Policy {
  private var iteration = 0

  val totalIterations: Int =
    if (config.repetitions < 0) currentBenchmark.defaultRepetitions else config.repetitions

  def description = "Runs the benchmark for a fixed number of iterations."

  /** The current iteration.
   */
  def currentIteration: Int = iteration

  private[renaissance] def execute(): Try[Unit] = {
    while (iteration < totalIterations) {
      val name = currentBenchmark.name
      val g = currentBenchmark.mainGroup
      if (iteration == totalIterations - 1) {
        println(s"====== $name ($g), final iteration started ======")
      } else {
        println(s"====== $name ($g), iteration $iteration started ======")
      }
      val nanos = currentBenchmark.runIterationWithBeforeAndAfter(this, config)
      val millis = (nanos / 1000).toInt / 1000.0
      if (iteration == totalIterations - 1) {
        println(s"====== $name ($g), final iteration completed ($millis ms) ======")
      } else {
        println(s"====== $name ($g), iteration $iteration completed ($millis ms) ======")
      }
      iteration += 1
    }
    Success(())
  }
}

final class Dummy extends RenaissanceBenchmark {
  override def description: String = "A dummy benchmark, which does no work."

  protected override def runIteration(config: Config): Unit = {}
}

File: renaissance-core/src/main/java/org/renaissance/RenaissanceBenchmark.java
package org.renaissance;

import java.io.IOException;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.function.BiFunction;
import java.util.regex.Pattern;

public abstract class RenaissanceBenchmark {
  public static final String kebabCase(String camelCaseName) {
    // This functionality is duplicated in the kebabCase function of the build file.
    Pattern pattern = Pattern.compile("([A-Za-z])([A-Z])");
    String result = camelCaseName;
    do {
      String last = result;
      result = pattern.matcher(result).replaceFirst("$1-$2");
      if (last == result) break;
    } while (true);
    return result.toLowerCase();
  }

  private volatile Object blackHoleField;

  public final String name() {
    Class<?> benchClass = this.getClass();
    Benchmark.Name annotation = benchClass.getDeclaredAnnotation(Benchmark.Name.class);
    if (annotation != null) {
      return annotation.value();
    } else {
      String cn = benchClass.getSimpleName();
      String camelCaseName =
        (cn.charAt(cn.length() - 1) == '$') ? cn.substring(0, cn.length() - 1) : cn;
      return kebabCase(camelCaseName);
    }
  }


  public final String mainGroup() {
    Class<?> benchClass = this.getClass();
    Benchmark.Group annotation = benchClass.getDeclaredAnnotation(Benchmark.Group.class);
    if (annotation != null) {
      return annotation.value();
    } else {
      String groupPkg = getPackageRelativeTo(benchClass, Benchmark.class);
      return groupPkg.replaceAll("[.]", "-");
    }
  }

  private static String getPackageRelativeTo (Class<?> target, Class<?> base) {
    final String targetPkg = target.getPackage().getName();
    final String basePkg = base.getPackage().getName();
    if (targetPkg.startsWith(basePkg)) {
      return targetPkg.substring(basePkg.length() + 1);
    } else {
      return targetPkg;
    }
  }

  public int defaultRepetitions() {
    Class<?> benchClass = this.getClass();
    Benchmark.Repetitions annotation = benchClass.getDeclaredAnnotation(Benchmark.Repetitions.class);
    return (annotation != null) ? annotation.value() : 20;
  }

  public void setUpBeforeAll(Config c) {
  }

  public void tearDownAfterAll(Config c) {
  }

  public void beforeIteration(Config c) {
  }

  public void afterIteration(Config c) {
  }

  public final Throwable runBenchmark(Config config) {
    try {
      setUpBeforeAll(config);
      if (!Policy.factories.containsKey(config.policy())) {
        System.err.println("Unknown policy " + config.policy() + ".");
        System.exit(1);
      }
      BiFunction<RenaissanceBenchmark, Config, Policy> factory =
        Policy.factories.get(config.policy());
      Policy policy = factory.apply(this, config);
      for (Plugin plugin : config.plugins()) {
        plugin.onStart(policy);
      }
      try {
        policy.execute();
      } finally {
        for (Plugin plugin : config.plugins()) {
          plugin.onTermination(policy);
        }
      }
      return null;
    } catch (Throwable t) {
      return t;
    } finally {
      try {
        tearDownAfterAll(config);
      } catch (Throwable t) {
        System.err.println("Error during tear-down: " + t.getMessage());
        t.printStackTrace();
      }
    }
  }

  protected <T> void blackHole(T value) {
    blackHoleField = value;
  }

  /**
   * This method runs the functionality of the benchmark.
   */
  protected abstract BenchmarkResult runIteration(Config config);

  long runIterationWithBeforeAndAfter(Policy policy, Config config) {
    long unixTsBefore = System.currentTimeMillis();

    beforeIteration(config);

    for (Plugin plugin : config.plugins()) {
      plugin.beforeIteration(policy);
    }

    long start = System.nanoTime();

    BenchmarkResult result = runIteration(config);

    long end = System.nanoTime();
    long duration = end - start;

    for (Plugin plugin : config.plugins()) {
      plugin.afterIteration(policy, duration);
    }

    afterIteration(config);

    long unixTsAfter = System.currentTimeMillis();

    result.validate();

    for (ResultObserver observer : config.resultObservers()) {
      observer.onNewResult(name(), "nanos", duration);
      observer.onNewResult(name(), "unixts.before", unixTsBefore);
      observer.onNewResult(name(), "unixts.after", unixTsAfter);
    }

    return duration;
  }

  public static void deleteTempDir(Path dirPath) {
    try {
      deleteRecursively(dirPath);
    } catch (Throwable t) {
      System.err.println("Error removing temp directory! " + t.getMessage());
    }
  }

  private static void deleteRecursively(final Path dirPath) throws IOException {
    Files.walkFileTree(dirPath, new SimpleFileVisitor<Path>() {
      @Override
      public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
        return delete(file);
      }

      @Override
      public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
        return delete(dir);
      }

      private FileVisitResult delete(Path path) throws IOException {
        Files.delete(path);
        return FileVisitResult.CONTINUE;
      }
    });
  }

  public static Path generateTempDir(String name) {
    Path p = null;
    try {
      p = Files.createTempDirectory(Paths.get("."), name);
    } catch (IOException e) {
      System.err.println("Error creating temp directory! " + e.getMessage());
    }
    return p;
  }
}

